### PR TITLE
Add-Devise-authentication,-Users-model,-Add-tests

### PR DESCRIPTION
**Pull request by fh0enix**
Title: Adding Devise Authentication and Users Model with Test Cases

Description:

This pull request introduces essential changes to enhance the security and user management of our application. Here's a breakdown of what this pull request accomplishes:

Devise Authentication Integration: Devise authentication has been integrated into our application. This ensures a robust and reliable user authentication system, providing a secure gateway for our users to access the platform.
image

Users Model Implementation: A dedicated "User" model has been introduced. This model encapsulates essential user-related data and functionalities, paving the way for efficient user management.
image

Comprehensive Test Suite: To maintain the reliability of our application, an extensive suite of tests has been included. These tests rigorously examine the functionality of the authentication system, ensuring that our users' data remains protected and our features operate as expected.
image

The integration of Devise and the introduction of the Users model significantly improve our application's security and user management. The comprehensive test coverage guarantees the stability of these changes.

Please review these modifications thoroughly, paying special attention to the security aspects and the impact on user functionality. Your feedback and suggestions are highly appreciated to ensure the successful implementation of this critical update.

Reviewer:

@loqimean

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -1,3 +1,11 @@
-
 class ApplicationController < ActionController::Base
+  before_action :configure_permitted_parameters, if: :devise_controller?
+  #before_action :authenticate_user!
```
Comment (loqimean) [inline app/controllers/application_controller.rb]: видаляй зайве

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -1,3 +1,11 @@
...
+
+protected
+
+def configure_permitted_parameters
+  devise_parameter_sanitizer.permit(:sign_up, keys:[:first_name, :last_name, :user_name])
+  devise_parameter_sanitizer.permit(:account_update, keys: [:first_name, :last_name, :user_name])
```
Comment (loqimean) [inline app/controllers/application_controller.rb]: де відступи? кодстайл

File: app/models/user.rb [hunk 3]
```diff
@@ -0,0 +1,8 @@
+class User < ApplicationRecord
+  # Include default devise modules. Others available are:
+  # :confirmable, :lockable, :timeoutable, :trackable and :omniauthable
+  devise :database_authenticatable, :registerable,
+         :recoverable, :rememberable, :validatable
+
+validates :first_name, :last_name, :user_name, presence: true
```
Comment (loqimean) [inline app/models/user.rb]: де відступи?

File: app/views/layouts/_header.html.erb [hunk 4]
```diff
@@ -0,0 +1,31 @@
+<header class="p-4 dark:bg-gray-800 dark:text-gray-100 rounded-b-2xl">
+	<div class="container flex justify-between h-16 mx-auto">
+		<ul class="items-stretch hidden lg:flex">
+			<li class="flex">
+				<a><%= link_to "UPLOAD NEW BOOK", new_book_path, class: "flex items-center px-4 -mb-1 border-b-2 dark:border-transparent dark:text-cyan-400 dark:border-cyan-400" %></a>
+			</li>
+		</ul>
+		<a rel="noopener noreferrer" href="/books" aria-label="Back to homepage" class="flex items-center p-2 duration-75 delay-1000 animate-pulse">
+			<svg xmlns="http://www.w3.org/2000/svg" fill="currentColor" viewBox="0 0 32 32" class="w-8 h-8 dark:text-cyan-400">
```
Comment (loqimean) [inline app/views/layouts/_header.html.erb]: `inline_svg`

File: app/views/layouts/application.html.erb [hunk 5]
```diff
@@ -12,11 +12,14 @@
...
 
   <body class="bg-gray-400">
+    <%= render 'layouts/header' %>
+
 
     <div class="flash">
       <%= render 'layouts/alerts' %>
     </div>
 
+
```
Comment (loqimean) [inline app/views/layouts/application.html.erb]:
```suggestion
```

File: app/views/layouts/application.html.erb [hunk 6]
```diff
@@ -12,11 +12,14 @@
   </head>
 
   <body class="bg-gray-400">
+    <%= render 'layouts/header' %>
+
```
Comment (loqimean) [inline app/views/layouts/application.html.erb]:
```suggestion
```

File: app/views/layouts/_header.html.erb [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+				<%= link_to "Sign Up", new_user_registration_path, class: "self-center px-8 py-3 font-semibold rounded dark:bg-cyan-400 dark:text-gray-900" %>
+			<% end %>
+		</div>
+		<button class="p-4 lg:hidden">
+			<svg xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 24 24" stroke="currentColor" class="w-6 h-6 dark:text-gray-100">
+				<path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M4 6h16M4 12h16M4 18h16"></path>
+			</svg>
+		</button>
+	</div>
+</header>
```
Comment (loqimean) [inline app/views/layouts/_header.html.erb]: чому тут по 4 пробіли? якщо не більше. виправляй

File: app/views/layouts/_header.html.erb [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+					<%="Hello #{current_user.first_name}!"%>
+				</div>
+				<%= button_to "Log Out", destroy_user_session_path, method: :delete, class: "self-center px-8 py-3 font-semibold rounded dark:bg-amber-400 dark:text-gray-900" %>
+			<% else %>
+				<%= link_to "Sign in", user_session_path, class: "self-center px-8 py-3 rounded" %>
+				<%= link_to "Sign Up", new_user_registration_path, class: "self-center px-8 py-3 font-semibold rounded dark:bg-cyan-400 dark:text-gray-900" %>
+			<% end %>
+		</div>
+		<button class="p-4 lg:hidden">
+			<svg xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 24 24" stroke="currentColor" class="w-6 h-6 dark:text-gray-100">
```
Comment (loqimean) [inline app/views/layouts/_header.html.erb]: `inline_svg`

File: config/routes.rb [hunk 9]
```diff
@@ -1,12 +1,8 @@
-
 Rails.application.routes.draw do
+  devise_for :users
```
Comment (loqimean) [inline config/routes.rb]: тобі прям всі роути тре?

File: db/seeds.rb [hunk 10]
```diff
@@ -1,3 +1,3 @@
 require "factory_bot_rails"
 
-FactoryBot.create_list(:book, 100)
+FactoryBot.create_list(:book, 5)
```
Comment (loqimean) [inline db/seeds.rb:3]: а юзери?

File: spec/models/user_spec.rb [hunk 11]
```diff
@@ -0,0 +1,19 @@
+require 'rails_helper'
+
+RSpec.describe User, type: :model do
+  describe 'validations' do
+    it { should validate_presence_of(:email) }
```
Comment (loqimean) [inline spec/models/user_spec.rb]: це старий синтаксис. міняй

File: spec/models/user_spec.rb [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+    it { should validate_presence_of(:first_name) }
+    it { should validate_presence_of(:last_name) }
+    it { should validate_presence_of(:user_name) }
+  end
+
+  describe 'factory' do
+    it 'has a valid factory' do
+      expect(FactoryBot.create(:user)).to be_valid
+    end
+  end
```
Comment (loqimean) [inline spec/models/user_spec.rb]: це вверх і без класу юзай

File: spec/rails_helper.rb [hunk 13]
```diff
@@ -35,6 +34,7 @@
 RSpec.configure do |config|
   # Remove this line if you"re not using ActiveRecord or ActiveRecord fixtures
   config.fixture_path = Rails.root.join("spec/fixtures").to_s
+  config.include Devise::Test::ControllerHelpers, type: :controller
```
Comment (loqimean) [inline spec/rails_helper.rb:40]: не в той тип передав

File: spec/requests/users_spec.rb [hunk 14]
```diff
@@ -0,0 +1,50 @@
+require 'rails_helper'
+
+RSpec.describe 'User sign up', type: :request do
+  describe 'POST #create' do
+    context 'with valid attributes' do
+      it 'creates a new user' do
+        expect do
+          post user_registration_path, params: { user: FactoryBot.attributes_for(:user) }
```
Comment (loqimean) [inline spec/requests/users_spec.rb]: фекторі_бот клас дефайниться глобально, нема сенсу писати константу

**Review by loqimean — CHANGES_REQUESTED** [review]

File: app/controllers/application_controller.rb (hunk 1 above)
Comment (fh0enix) [reply app/controllers/application_controller.rb]: solved

File: app/models/user.rb (hunk 3 above)
Comment (fh0enix) [reply app/models/user.rb]: sorry
solved

File: app/views/layouts/application.html.erb (hunk 6 above)
Comment (fh0enix) [reply app/views/layouts/application.html.erb]: solved

File: app/views/layouts/application.html.erb (hunk 5 above)
Comment (fh0enix) [reply app/views/layouts/application.html.erb]: solved

File: config/routes.rb (hunk 9 above)
Comment (fh0enix) [reply config/routes.rb]: solved

File: db/seeds.rb (hunk 10 above)
Comment (fh0enix) [reply db/seeds.rb:3]: solved

File: spec/requests/users_spec.rb (hunk 14 above)
Comment (fh0enix) [reply spec/requests/users_spec.rb]: solved

File: spec/models/user_spec.rb (hunk 12 above)
Comment (fh0enix) [reply spec/models/user_spec.rb]: deleted it

File: spec/models/user_spec.rb (hunk 11 above)
Comment (fh0enix) [reply spec/models/user_spec.rb]: solved

File: spec/rails_helper.rb (hunk 13 above)
Comment (fh0enix) [reply spec/rails_helper.rb:40]: solved?

File: app/views/layouts/_header.html.erb (hunk 7 above)
Comment (fh0enix) [reply app/views/layouts/_header.html.erb]: The code містить пробіли, які відповідають HTML-відступам для кращої читабельності коду

File: app/views/layouts/_header.html.erb (hunk 8 above)
Comment (fh0enix) [reply app/views/layouts/_header.html.erb]: solved

File: app/views/layouts/_header.html.erb (hunk 4 above)
Comment (fh0enix) [reply app/views/layouts/_header.html.erb]: solved

File: config/routes.rb [hunk 15]
```diff
@@ -1,12 +1,8 @@
...
+  devise_for :users, only: [:sessions, :registrations]
   resources :books
-  # Define your application routes per the DSL in https://guides.rubyonrails.org/routing.html
 
-  # Reveal health status on /up that returns 200 if the app boots with no exceptions, otherwise 500.
-  # Can be used by load balancers and uptime monitors to verify that the app is live.
   get "up" => "rails/health#show", as: :rails_health_check
 
-  # Defines the root path route ("/")
   root "books#index"
```
Comment (loqimean) [inline config/routes.rb]: на верх перед девайсом

File: config/routes.rb [hunk 16]
```diff
@@ -1,12 +1,8 @@
-
 Rails.application.routes.draw do
+  devise_for :users, only: [:sessions, :registrations]
```
Comment (loqimean) [inline config/routes.rb:2]:
```suggestion
  devise_for :users, only: [:sessions, :registrations]

```

File: spec/models/user_spec.rb [hunk 17]
```diff
@@ -0,0 +1,11 @@
+require 'rails_helper'
+
+RSpec.describe User, type: :model do
+    it { is_expected.to validate_presence_of(:email) }
+    it { is_expected.to validate_uniqueness_of(:email).case_insensitive }
+    it { is_expected.to validate_presence_of(:password) }
+    it { is_expected.to validate_length_of(:password).is_at_least(6) }
+    it { is_expected.to validate_presence_of(:first_name) }
+    it { is_expected.to validate_presence_of(:last_name) }
+    it { is_expected.to validate_presence_of(:user_name) }
```
Comment (loqimean) [inline spec/models/user_spec.rb]: два пробіли

File: spec/requests/users_spec.rb (hunk 14 above)
Comment (loqimean) [reply spec/requests/users_spec.rb]: не солвед)

File: app/controllers/application_controller.rb [hunk 18]
```diff
@@ -1,3 +1,11 @@
...
 class ApplicationController < ActionController::Base
+  before_action :configure_permitted_parameters, if: :devise_controller?
+
+protected
+
+def configure_permitted_parameters
+  devise_parameter_sanitizer.permit(:sign_up, keys:[:first_name, :last_name, :user_name])
+
+  devise_parameter_sanitizer.permit(:account_update, keys: [:first_name, :last_name, :user_name])
+end
```
Comment (loqimean) [inline app/controllers/application_controller.rb]: два пробіли

File: app/views/layouts/_header.html.erb [hunk 19]
```diff
@@ -0,0 +1,28 @@
...
+		</a>
+		<ul class="items-stretch hidden lg:flex">
+			<li class="flex">
+				<% if current_user %>
+					<a><%= link_to "UPLOAD NEW BOOK", new_book_path, class: "duration-75 delay-1000 animate-pulse flex items-center px-4 -mb-1 border-b-2 dark:border-transparent dark:text-cyan-400 dark:border-cyan-400" %></a>
+				<% end %>
+			</li>
+		</ul>
+		<div class="items-center flex-shrink-0 hidden lg:flex">
+			<% if current_user %>
```
Comment (loqimean) [inline app/views/layouts/_header.html.erb]: код стайл

**Review by loqimean — CHANGES_REQUESTED** [review]

File: app/controllers/application_controller.rb (hunk 18 above)
Comment (fh0enix) [reply app/controllers/application_controller.rb]: solved

File: app/views/layouts/_header.html.erb (hunk 19 above)
Comment (fh0enix) [reply app/views/layouts/_header.html.erb]: solved

File: config/routes.rb (hunk 15 above)
Comment (fh0enix) [reply config/routes.rb]: solved

File: config/routes.rb (hunk 16 above)
Comment (fh0enix) [reply config/routes.rb:2]: solved

File: spec/models/user_spec.rb (hunk 17 above)
Comment (fh0enix) [reply spec/models/user_spec.rb]: solved

File: app/controllers/application_controller.rb (hunk 2 above)
Comment (fh0enix) [reply app/controllers/application_controller.rb]: solved

File: spec/requests/users_spec.rb (hunk 14 above)
Comment (fh0enix) [reply spec/requests/users_spec.rb]: now I have desided it

File: app/views/books/_show_book.html.erb [hunk 20]
```diff
@@ -0,0 +1,23 @@
...
+		<div class="space-y-2">
+			<img src="https://source.unsplash.com/random/480x360/?4" alt="" class="book-cover">
+			<div class="flex items-center text-xs">
+				<span>ISBN:<%= book.isbn %></span>
+			</div>
+		</div>
+		<div class="space-y-2">
+			<a rel="noopener noreferrer" href="#" class="block">
+				<h4 class="text-xl font-semibold dark:dark:text-cyan-400"><%= book.author %></h4>
+				<h3 class="text-xl font-semibold dark:dark:text-gray-100"><%= book.title %></h3>
```
Comment (loqimean) [inline app/views/books/_show_book.html.erb]: ну що це за приколи з табами по 8 пробілів?))

File: app/views/layouts/_header.html.erb [hunk 21]
```diff
@@ -0,0 +1,25 @@
...
+				<div class="self-center px-8 py-3 rounded">
+					<ul class="items-stretch hidden lg:flex">
+						<li class="flex">
+								<a><%= link_to "Hello #{current_user.first_name}! UPLOAD NEW BOOK", new_book_path, class: "btn-hello" %></a>
+						</li>
+					</ul>
+				</div>
+				<%= button_to "Log Out", destroy_user_session_path, method: :delete, class: "self-center px-8 py-3 font-semibold rounded dark:bg-amber-400 dark:text-gray-900" %>
```
Comment (loqimean) [inline app/views/layouts/_header.html.erb]: та ж історія

File: config/routes.rb [hunk 22]
```diff
@@ -1,12 +1,9 @@
-
 Rails.application.routes.draw do
+  root "books#index"
+
+  devise_for :users, only: [:sessions, :registrations]
```
Comment (loqimean) [inline config/routes.rb]:
```suggestion
  devise_for :users, only: [:sessions, :registrations]

  root "books#index"
```

**Review by loqimean — CHANGES_REQUESTED** [review]

File: config/routes.rb (hunk 22 above)
Comment (fh0enix) [reply config/routes.rb]: solved

File: app/views/layouts/_header.html.erb [hunk 23]
```diff
@@ -0,0 +1,25 @@
+<nav class="sticky top-0 p-4 dark:bg-gray-900 dark:text-gray-100 rounded-b-2xl">
+  <div class="container flex justify-between h-16 mx-auto">
+    <a href="/books" class="flex items-center p-2">
+      <%= inline_svg "book.svg", class: "w-8 h-8 text-cyan-400" %>
+    </a>
+    <div class="items-center flex-shrink-0 hidden lg:flex">
+      <% if current_user %>
```
Comment (loqimean) [inline app/views/layouts/_header.html.erb]:
```suggestion
      <% if current_user.present? %>
```

File: app/views/layouts/application.html.erb (hunk 6 above)
Comment (loqimean) [reply app/views/layouts/application.html.erb]: не солвед)

**Review by loqimean — CHANGES_REQUESTED** [review]

File: app/views/layouts/application.html.erb (hunk 6 above)
Comment (fh0enix) [reply app/views/layouts/application.html.erb]: солвед)

File: app/views/layouts/_header.html.erb (hunk 23 above)
Comment (fh0enix) [reply app/views/layouts/_header.html.erb]: solved

**Review by loqimean — APPROVED** [review]

**Pull request merged**